### ISSUE TITLE
chat history problem in blender_90m

### ISSUE THREAD

**Issue by Hilal-Urun** [facebookresearch/ParlAI#3421]
Hi , I'm working on blender 90m model and I just use spesipic persona for agent. In torch_agent.py file, I changed the --history_size as lenght of the history string but when I'm try to chat with agent, for example 10 chat logs later when I ask to something of the things we talked about, it can not answer and also sometimes, it asks the questions asked before. Is there any solution to help my problem, I examined the History func several times and tried some solutions but it didn't work.
Thanks for your time.

**Comment by stephenroller** [comment]
The model can only ever see 128 tokens back, or about 6 turns. It's just a hard limitation of the model.

That said, you can use `--beam-block-full-context true` to block it from saying the _exact_ same thing for infinite turns, but it only helps a little bit.

**Comment by Hilal-Urun** [comment]
I got it thanks a lot. Is it same for all other models?

**Comment by stephenroller** [comment]
yes pretty much

**Comment by Hilal-Urun** [comment]
![image](https://user-images.githubusercontent.com/37075702/108838866-47f2ff00-75e5-11eb-93cc-0aedd4e8008d.png)
Hi again. I just tryied to fix input as ; 
![image](https://user-images.githubusercontent.com/37075702/108838921-63f6a080-75e5-11eb-9ed6-0f2b31364270.png)
and I gave the spesific persona like bot_persona = "\n".join(["your persona: my name is dora. i like to travel. i am 24 years old. i am single."]) . But there is a problem here which is in the first picture. How can I handle with that.  Could you give an idea ? @stephenroller

**Comment by stephenroller** [comment]
The persona info should look like this (with new lines):

```
your persona: my name is dora.
your persona: i like to travel.
your persona: i am 24 years old.
your persona: i am single.
```

I can't tell from context, but you should only give the persona on the _very_ first turn. After that, it's already in the context.

Another small tip, which doesn't matter for the 90M but does for others, is that you don't want to lowercase the history.

FYI, as long as you're using the 90M, you may be interested in [this helper method](https://github.com/facebookresearch/ParlAI/blob/7293ffe7b3d6acd54d22e96411bb8a535769052c/parlai/utils/strings.py#L13) that fixes much of the punctuation/capitalization.

**Comment by Hilal-Urun** [comment]
thanks a lot for information, now i am using the normalize_reply(). Yes I tried to give persona at the first stage but after a while persona is changes like name or jobs are changes before the episode done.

**Comment by stephenroller** [comment]
The models are limited to about ~6 turns of memory. Just a limitation of the state of the art. The 90M model should have a memory of more like 12-18 turns but use it less effectively.

**Comment by Hilal-Urun** [comment]
okey i got it, thanks a lot @stephenroller then, can it be fixed on the model?

**Comment by stephenroller** [comment]
It requires pre-training a new model from scratch with more GPUs or a new architecture.
